### PR TITLE
archivebox: fix build and add singlefile

### LINKED ISSUE #358580: Build failure: archivebox on NixOS 24.11
## Steps To Reproduce

archivebox doesn't build on NixOS 24.11.

I've fixed issues with `django-extensions` patches and reverted to Python 3.11 (`django` needs `distutils` which was removed in 3.12), but the build fails with `pdm.backend` not being available.

Which is odd as `pdm-backend` is already in `nativeBuildInputs`! I've tried putting it in `build-systems` too but no luck, either.

```patch
diff --git a/pkgs/by-name/ar/archivebox/package.nix b/pkgs/by-name/ar/archivebox/package.nix
index 4344c6e7e91d..9e8d760c978a 100644
--- a/pkgs/by-name/ar/archivebox/package.nix
+++ b/pkgs/by-name/ar/archivebox/package.nix
@@ -44,6 +44,8 @@ let
           rev = "e43f383dae3a35237e42f6acfe1207a8e7e7bdf5";
           hash = "sha256-NAMa78KhAuoJfp0Cb0Codz84sRfRQ1JhSLNYRI4GBPM=";
         };
+        patches = [ ];
+        postPatch = '''';
 
         # possibly a real issue, but that version is not supported anymore
         doCheck = false;
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 8d7b98e467f9..f9ca3954ee37 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -11679,6 +11679,10 @@ with pkgs;
 
   appdaemon = callPackage ../servers/home-assistant/appdaemon.nix { };
 
+  archivebox = callPackage ../by-name/ar/archivebox/package.nix {
+    python3 = python311;
+  };
+
   cassandra_3_0 = callPackage ../servers/nosql/cassandra/3.0.nix {
     jre = jre8; # TODO: remove override https://github.com/NixOS/nixpkgs/pull/89731
     python = python2;
```

```bash
NIXPKGS_ALLOW_INSECURE=1 nix build .#archivebox -L --impure
```

## Build log

<details>

<summary>Build Log</summary>

```
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/sylb1w159ws447905220wxcbl87c9777-archivebox-0.7.2.tar.gz
source root is archivebox-0.7.2
setting SOURCE_DATE_EPOCH to timestamp 1704424882 of file archivebox-0.7.2/pyproject.toml
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing pypaBuildPhase
Setting PDM_BUILD_SCM_VERSION to 0.7.2
Creating a wheel...
pypa build flags: --no-isolation --outdir dist/ --wheel
* Getting build dependencies for wheel...

Traceback (most recent call last):
  File "/nix/store/mipybsy9x06av6j90qs7c5wsb0nc06h3-python3.11-pyproject-hooks-1.2.0/lib/python3.11/site-packages/pyproject_hooks/_impl.py", line 402, in _call_hook
    raise BackendUnavailable(
pyproject_hooks._impl.BackendUnavailable: Cannot import 'pdm.backend'

ERROR Backend 'pdm.backend' is not available.
```

</details>

## Notify maintainers

@siraben @viraptor 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by james-atkins**
Fixes #358580.

Override any django-extensions patches.
~~Downgrade to Python 3.11 as django needs distutils.~~

Avoid a scoping error with yt-dlp. If `pkgs.yt-dlp` is in dependencies instead of `python.pkgs.yt-dlp`, there is
a cryptic error about 'pdm.backend' not being available. Change to just using `python.pkgs.yt-dlp` everywhere.

Also update `buildPythonApplication` to use `build-system` and `dependencies`. 

Add SingleFile support too.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by dotlambda — COMMENTED** [review]
Why not remove `yt-dlp` from the inputs instead?
Also, the commit message doesn't make sense since we're merging this into master first.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -11609,6 +11609,10 @@ with pkgs;
 
   appdaemon = callPackage ../servers/home-assistant/appdaemon.nix { };
 
+  archivebox = callPackage ../by-name/ar/archivebox/package.nix {
+    python3 = python311;
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:11613]:
```suggestion
```
Add `distutils` to `build-system` (I assume) instead.

**Comment by dotlambda** [comment]
> We need `pkgs.yt-dlp` for the environment variable later in the code so we can't remove it.

Use `python.pkgs.yt-dlp` for that.

**Comment by dotlambda** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358588`

---
### `x86_64-linux`

**Review by dotlambda — APPROVED** [review]

**Comment by siraben** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358588`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>archivebox</li>
    <li>archivebox.dist</li>
  </ul>
</details>

**Comment by dotlambda** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358588`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>archivebox</li>
    <li>archivebox.dist</li>
  </ul>
</details>

File: pkgs/by-name/ar/archivebox/package.nix [hunk 2]
```diff
@@ -34,6 +34,7 @@ let
             "CVE-2022-28346"
           ];
         };
+        dependencies = (old.dependencies or [ ]) ++ (lib.optionals (python.pythonAtLeast "3.12") [ python.pkgs.distutils ]);
```
Comment (dotlambda) [inline pkgs/by-name/ar/archivebox/package.nix:37]: Why is distutils in dependencies rather than build-system?

Comment (dotlambda) [inline pkgs/by-name/ar/archivebox/package.nix:37]:
```suggestion
        dependencies = (old.dependencies or [ ]) ++ (lib.optionals (self.pythonAtLeast "3.12") [ self.distutils ]);
```

**Pull request merged**